File: Linked_list/linked_list.py
class Node:
    def __init__ (self, value = None, next = None):
        self.next = next;
        self.value = value;

class Linked_list:
    def __init__ (self):
        self.head = None

    def insert_at_beginning (self, data):
        node = Node(data, self.head)
        self.head = node

    def print_list (self):
        if self.head is None:
            print('Linked List is empty')
            return
        
        itr = self.head
        listr = ''
        while itr:
            listr += str(itr.value) + '--'
            itr = itr.next
        print(listr)

    def insert_at_end(self, data):
        if self.head is None:
            self.head = Node(data, None)
            return
    
        itr = self.head
        while itr.next:
            itr = itr.next

        itr.next = Node(data, None)

    def insert_value (self, data_list):
        self.head = None
        for data in data_list:
            self.insert_at_end(data)

    def get_length (self):
        count = 0
        itr = self.head
        while itr:
            count += 1
            itr = itr.next

        return count
    
    def remove_at (self, index):
        if index < 0 or index >= self.get_length():
            raise Exception("Invalid Value of index")

        if index == 0:
            self.head = self.head.next
            return
        
        count = 0
        itr = self.head
        while itr:
            if count == index - 1:
                itr.next = itr.next.next
                break

            count += 1
            itr = itr.next


    def insert_at(self, index, data):
        if index < 0 or index > self.get_length():
            raise Exception("Invalid Value of index")
        if index == 0:
            self.insert_at_beginning(data)
            return
        
        count = 0
        itr = self.head
        while itr:
            if count == index - 1:
                node = Node(data, itr.next)
                itr.next = node
                return

            count += 1
            itr = itr.next

        
if __name__ == '__main__':
    li = Linked_list()
    li.insert_at_beginning(1)
    li.insert_at_beginning(2)
    li.insert_at_beginning(3)
    li.insert_at_beginning(4)
    li.insert_at_end(9)
    li.insert_at_beginning(0)
    li.print_list()
            
    li2 = Linked_list()
    data_list = ['Orange', 'water', 'cocaca', 'banana']
    li2.insert_value(data_list)
    li2.print_list()
    print("li2 length: ", li2.get_length())
    li2.remove_at(2)
    print("li2 after remove node")
    li2.print_list()
    print("li2 after insert new value to a specific Aindex")
    li2.insert_at(1, "cheese")
    li2.print_list()

File: Linked_list/linked_list.cc
#include <iostream>

class Node {
public:
    int value;
    Node* next;

    Node(int data) : value(data), next(nullptr) {}
};

class Linked_list {
private:
    Node* head;

public:
    Linked_list() : head(nullptr) {}

    void insert_at_beginning(int data) {
        Node* newNode = new Node(data);
        newNode->next = head;
        head = newNode;
    }

    void print_list() {
        if (head == nullptr) {
            std::cout << "Linked List is empty" << std::endl;
            return;
        }

        Node* itr = head;
        while (itr != nullptr) {
            std::cout << itr->value << "--";
            itr = itr->next;
        }
        std::cout << std::endl;
    }

    void insert_at_end(int data) {
        if (head == nullptr) {
            head = new Node(data);
            return;
        }

        Node* itr = head;
        while (itr->next != nullptr) {
            itr = itr->next;
        }
        itr->next = new Node(data);
    }

    void insert_value(int* dataList, int size) {
        head = nullptr;
        for (int i = 0; i < size; ++i) {
            insert_at_end(dataList[i]);
        }
    }

    int get_length() {
        int count = 0;
        Node* itr = head;
        while (itr != nullptr) {
            count++;
            itr = itr->next;
        }
        return count;
    }

    void remove_at(int index) {
        if (index < 0 || index >= get_length()) {
            throw std::invalid_argument("Invalid index");
        }

        if (index == 0) {
            Node* temp = head;
            head = head->next;
            delete temp;
            return;
        }

        int count = 0;
        Node* itr = head;
        while (itr != nullptr) {
            if (count == index - 1) {
                Node* temp = itr->next;
                itr->next = itr->next->next;
                delete temp;
                break;
            }
            count++;
            itr = itr->next;
        }
    }

    void insert_at(int index, int data) {
        if (index < 0 || index > get_length()) {
            throw std::invalid_argument("Invalid index");
        }

        if (index == 0) {
            insert_at_beginning(data);
            return;
        }

        int count = 0;
        Node* itr = head;
        while (itr != nullptr) {
            if (count == index - 1) {
                Node* newNode = new Node(data);
                newNode->next = itr->next;
                itr->next = newNode;
                break;
            }
            count++;
            itr = itr->next;
        }
    }

    ~Linked_list() {
        Node* temp;
        while (head != nullptr) {
            temp = head;
            head = head->next;
            delete temp;
        }
    }
};

int main() {
    Linked_list li;
    li.insert_at_beginning(1);
    li.insert_at_beginning(2);
    li.insert_at_beginning(3);
    li.insert_at_beginning(4);
    li.insert_at_end(9);
    li.insert_at_beginning(0);
    li.print_list();
    
    Linked_list li2;
    int data[] = {2, 4, 6, 8};
    li2.insert_value(data, sizeof(data) / sizeof(data[0]));
    li2.print_list();
    std::cout << "li2 length: " << li2.get_length() << std::endl;
    li2.remove_at(2);
    std::cout << "li2 after remove node" << std::endl;
    li2.print_list();
    std::cout << "li2 after insert new value to a specific index" << std::endl;
    li2.insert_at(1, 5);
    li2.print_list();

    return 0;
}
